### ISSUE TITLE
Add TCP Health Check using SYN, SYN-ACK, and RST packets

### ISSUE THREAD

**Issue by eos175** [traefik/traefik#10794]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

I have implemented a new TCP health check functionality using the `tcp-shaker` library (https://github.com/tevino/tcp-shaker). This improvement uses SYN, SYN-ACK, and RST packets to perform more accurate health checks for TCP services.

You can review the implementation on my branch [here](https://github.com/eos175/traefik/tree/dev).

Would love to get feedback and discuss how this can be integrated into the project.

Thanks!

**Comment by eos175** [comment]
Some related issues are #5598 and #1657.

To use it, the procedure is as follows:

```yaml
services:
  example_tcp_service:
    loadBalancer:
      servers:
        - address: 10.42.64.4:7445
        - address: 10.42.0.1:7445
        - address: 127.0.0.1:8448
      healthCheck:
        path: /health
        interval: "10s"
        timeout: "3s"
        mode: tcp
```

Note that the `path` is ignored due to the nature of the TCP protocol.

Additionally, it would be interesting to consider a `mode: passive`, which enables passive checks based on observing communication with back-end nodes. However, this is not as straightforward.

**Comment by leonidas-o** [comment]
@nmengin what do you mean by "we’re unsure about the use case"? I mean it's a tcp health check, which is really needed in such a product. Just look at #5598 there you will see, they are asking for it.
Another example and real use case. You want to install OpenShift with traefik as a load balancer:

https://docs.okd.io/4.16/installing/installing_bare_metal/installing-bare-metal-network-customizations.html#installation-load-balancing-user-infra-example_installing-bare-metal-network-customizations

They use HAProxy in the example, as you can see `mode: tcp and check /readyz`, so there is a need for a tcp health check at leas if you want to setup openshift:
```bash
listen api-server-6443 
  bind *:6443
  mode tcp
  option  httpchk GET /readyz HTTP/1.0
  ...
```
I can't say anything about @eos175 's implementation, but I (and I'm pretty sure a lot of others) would really appreciate if you would accept that. Of course if the implementation quality is up to your defined standards.

**Comment by nmengin** [comment]
Hello @leonidas-o,

Thank you for your feedback.

> @nmengin what do you mean by "we’re unsure about the use case"?

I meant that, with the other maintainers, we're unsure that this feature is expected by a lot of people in the community.
We prefer to collect feedback before accepting it as an enhancement in Traefik.

**Comment by jhvst** [comment]
This proposal is just what I am looking for!

It covers these two cases:

- proxying a HTTPS server from downstream with TLS passthrough
- reundant TCP services

The first case is common in environments in which you have your own certificates that you must present to upstream. This is common in non-cloud environments, which host public services, but use a proxy to mask the origin IP source.

I am looking to migrate to traefik to replace Caddy and nginx, which I have to both use currently for the following reasons:

1. Caddy does not support TCP proxying
2. nginx requires a Plus subscription for monitoring TCP endpoint metrics via Prometheus

The proposal in this thread would provide a feature which would make traefik a one-stop solution for my multiple use-cases.

**Comment by dfrt82** [comment]
I second that, I definitely need this feature. Trying traefik at the moment with a very simple web application that uses websockets. A working http loadbalancing is of no use here as long as there is no way to automatically skip dead servers within TCP loadbalancing as well, otherwise traefik tries to establish websocket connections to the dead servers.

Also I have no idea why the use-case is questioned again, as @leonidas-o pointed out, #5598 should be enough evidence as you already asked for community support.

**Comment by eos175** [comment]
If anyone tests on the master [branch](https://github.com/eos175/traefik/tree/master), it includes changes up to version 3.1.0. So far, everything is working well, and there's no longer a 30s timeout delay.

![image](https://github.com/user-attachments/assets/1866f095-a88d-4da3-8806-5710f0c9fa30)

**Comment by nmengin** [comment]
Hello there,

Thank you for the feedback, we have set the status to enhancement.
However, this would not make it to our roadmap as we are focused elsewhere. If a community member would like to build it, let us know, and we will work with you to ensure you have all the information needed to merge it.

We prefer to work with our community members at the beginning of the design process to ensure that we are aligned and can move quickly with the review and merge process. Let us know here or create a PR before you start, and we will work with you there.

Don’t forget to check out the[ contributor docs](https://github.com/traefik/contributors-guide/blob/master/pr_guidelines.md) and link the PR to this issue.

**Comment by leonidas-o** [comment]
@nmengin feels like we are going in circles. I don't know am I missing something? 
1. @eos175 initial post (https://github.com/traefik/traefik/issues/10794#issue-2342848171) , the first post at the top clearly says: "I have implemented a new TCP health check...". So what is missing, a PR? @eos175 can you submit a PR if not already done?

2. I saw that @ddtmachado submitted a PR for a tcp health check: https://github.com/traefik/traefik/pull/11238

The big question is, what is the difference between the two implementations (@eos175 and @ddtmachado ) in terms of stability, security, performance etc.? Maybe we can even use some synergetic effects and get the best solution here?

**Comment by eos175** [comment]
Hi @nmengin, 

Thank you for your input, and I appreciate the opportunity to clarify. 

I haven’t submitted a PR yet because I’ve never contributed to Traefik before, and I’m unsure if my implementation meets Traefik’s standards. I wanted to gather feedback before formalizing the contribution. 

The main difference between my implementation and the one by @ddtmachado is the approach to sending `SYN` and `SYN-ACK` packets. My implementation leverages the `github.com/tevino/tcp-shaker` library, which allows sending these packets directly to the kernel without passing the request to the application. As a result, these connection attempts don’t appear in the logs, which I believe is an advantage in terms of cleaner operation and reduced noise in monitoring.  

Initially, my implementation performed a regular TCP connection, as seen in [this commit](https://github.com/eos175/traefik/commit/b80ce22626cb1b462a43b0d56f21b7a9533713e5#diff-2c565121f665a1bc64eef7d8aeb4c3c406473167bb5fb93426755d18c98fa204L173 ). However, I moved away from that approach because it closely resembles what I believe @ddtmachado’s implementation does — opening a full TCP connection and waiting for a response.  

From my perspective, this approach is not ideal. TCP is typically used as a transport layer for higher-level protocols, so expecting a meaningful response at this level doesn’t align with its intended use. For instance, if a response is expected, it should ideally use a protocol like HTTP. This behavior aligns more with what HAProxy does, which focuses on using TCP for its core purpose without conflating it with application-level expectations.

I’d be happy to collaborate further to refine the solution and ensure the best approach for Traefik. Let me know how you’d like to proceed!

**Comment by leonidas-o** [comment]
I'm going to summarise things up, as well as add my thoughts. If one of my assumptions is wrong, please correct me. Also I'm not a maintainer or involved in traefik in any way, but I would like to see a TCP Health Check feature in traefik. Therefore trying to help with this a bit.
It looks like, we are talking about these two implementations:

## Approaches
@ddtmachado 
1. Establishing a complete tcp connection: `Syn, Syn/Ack, Ack`
- What happens with the established connection? Is it kept open, for how long, some data send, immediately closed (client and server send Fin/Ack, Fin/Ack)?
- So basically a full three way handshake for the connection and a two way handshake for the termination?
- Are you using any external go package for that/ external dependencies?

@eos175 
2. You are using the `tcp-shaker` package to only do a: `Syn, Syn/Ack, Rst`

- Which means, as soon as the server replies with the Syn/Ack, the connection is terminated via `Rst`?
- `Rst` is definitely send after Syn/Ack? Or could it be, that the package simply does a `Syn, Syn/Ack` without sending a `Rst` which would result in what, a server waiting - for the length of a timeout - until the `Rst`/ usually `Ack` arrives? Which would basically lead to a wasted connection.


## Information/ Collection of statements

A comment states that:
```
"... I got from some expert that it is usually done for improve performance since it will have less handshake than normal connection. ..."
```
> https://serverfault.com/questions/884982/is-rst-ack-expected-in-a-tcp-health-check


Fortinet seems to do `Syn, Syn/Ack, Rst` (although they just write `Syn, Syn/Ack`)
```
Layer 4 TCP and UDP health checks are "liveness" checks. At Layer 4, FortiADC sends health check probes and attempts to connect to a specific TCP or UDP port. A TCP SYN request to port 80, for example, would expect a TCP SYN ACK response in return. If the response is not received within the configured timeouts and intervals, port 80 would be marked DOWN for that server, or server instance.
```
> https://help.fortinet.com/fadc/4-2-2/html-e/Content/Health_Checks/TLHC_L4_HC.htm


Tencent cloud is doing `Syn, Syn/Ack, Rst`
>https://www.tencentcloud.com/document/product/214/6097#tcp-.E5.81.A5.E5.BA.B7.E6.A3.80.E6.9F.A5


Envoy, not really sure as “connect only” TCP health checking could mean both.
```
If “receive” is an empty array, Envoy will perform “connect only” TCP health checking. During each cycle, Envoy will attempt to connect to the upstream host, and consider it a success if the connection succeeds. A new connection is created for each health check cycle.
```
> https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_hc


HaProxy uses `Syn, Syn/Ack, Rst`
```
these sequences of packets including a RST are completely normal. This is the way that haproxy uses to do health checks efficiently. As soon as haproxy has discovered that the endpoint is up, there is no point in wasting any further resources at either end. It turns out that using TCP RST is the most efficient way for kernels at both ends of the connection to finish their conversation and free up those resources.
```
> https://discourse.haproxy.org/t/connection-reset-seen-every-2-sec-haproxy/2156/5

Radware Alteon
```
The TCP handshake flow is SYN, SYN/ACK and RST like the example of packet captures below shows.
```
> https://support.radware.com/app/answers/answer_view/a_id/1028668/~/alteon-tcp-health-check-without-sending-ack


## Summary
It makes sense to say that `Syn, Syn/Ack, Rst` is the more efficient implementation, as it terminates immediately, no waiting on timeouts, no connections kept open for longer than needed. Also it looks like, others use exactly that approach (see link collection above).
Me personally, I don't like external dependencies and would say, avoid them as much as possible. Could `Syn, Syn/Ack, Rst` be achieved with go standard libraries (without a lot of extra/ duplicate work)?

@nmengin what are your thoughts on this? Which one of these two approaches would you/ other maintainers prefer and why?

**Comment by ddtmachado** [comment]
@leonidas-o your summary seems correct.

I opted for the full cycle to be able to support data in and out as a valid check, in case you run a custom protocol or some legacy server. I'm not using any external dependencies and tried to reuse most of what Traefik already have, for example I just pushed the config with TLS support using the servers transport config from Traefik.

IMO we could go with a Syn, Syn/Ack, Rst **if** its possible to introduce that within the Traefik's dialer manager, or somewhere else on the TCP transport level config, to avoid using external deps and creating a different dial flow just for the healthcheck. I'll wait for @nmengin to jump in on that.

BTW, here is a quick update on what's missing in my branch at the moment:

* mTLS setup and tests
* TLS test is broken (will look into it soon)
* Documentation

**Comment by leonidas-o** [comment]
@ddtmachado sound plausible, and I like your implementation choice keeping it in the traefik domain, so no external dependencies.

Okay, what about a little logic where you can either use a full cycle OR the `Syn, Syn/Ack, Rst` approach, does that make sense? 
So a user would have the option to choose the data in and out as a valid check, custom protocol, legacy server OR the performance oriented way.

**Comment by eos175** [comment]
`tcp-shaker` does exactly what HAProxy does:  

```
1. SYN  
2. SYN-ACK  
3. RST  
```  

I don’t understand why adding `tcp-shaker` as a dependency would be an issue, especially since it only depends on:  

- `github.com/pkg/errors`  
- `golang.org/x/sys`  

Both of which are already used by Traefik.  

I’m reusing the HTTP check logic but ignoring the path, as it’s not needed in this case. I think this approach is much lighter since there’s no need to worry about custom protocols, TLS, etc. After all, the goal here is simply to determine whether the TCP server can accept requests.

**Comment by leonidas-o** [comment]
As I said, I'm not a contributor, so I speak for myself. Maybe the traefik team has other thoughts on that.
And based on my research, I personally like `Syn, Syn/Ack, Rst` better and I would use it for my use case, but I also see, that it could live next to the full cycle method. Therefore my question, could you and @ddtmachado find a way to incorporate both and the user either activates his or your implementation?
It's good that tcp-shaker itself uses packages already imported in traefik and does not add other external dependencies.
As for the possible issues with external dependencies, I ask myself, is the community support behind tcp-shaker strong enough or could it get abandoned, could it introduce any risks, bad code etc. What happens if it gets abandoned, will the `Syn, Syn/Ack, Rst` feature be removed or will the team fork tcp-shaker and keep it alive, do they have capacities for that, etc. So many questions which come up when introducing external dependencies in general. That's something only the team can answer.

**Comment by mdklapwijk** [comment]
> [@ddtmachado](https://github.com/ddtmachado) sound plausible, and I like your implementation choice keeping it in the traefik domain, so no external dependencies.
> 
> Okay, what about a little logic where you can either use a full cycle OR the `Syn, Syn/Ack, Rst` approach, does that make sense? So a user would have the option to choose the data in and out as a valid check, custom protocol, legacy server OR the performance oriented way.

It would be great if the check supports a boolean for the lightweight "is port open" check and the full "is service answering" check. I think the former should be the default as it is the expected behaviour of many other products. The latter is very useful for services that sometimes keep accepting connections but stopped working correctly (e.g. forking issues after the initial connection), so it would be nice to be able to check for a triggered response or even only wait for the TLS-handshake to be finished correctly, with or without certificate checking.

One could also opt for a mix of the 2; only do full check every 10 checks or something like that.

**Comment by supergoudvis116** [comment]
@eos175 I really like how you did the health check, for us the TCP part is not really required but I decided to make changes on your PR to make it useful for our scenario. 

https://github.com/eos175/traefik/pull/1

In PostgreSQL HA clusters managed by Patroni, only the current leader is writable, and its status is exposed via a local HTTP API (returning HTTP 200 for the leader, 503 for replicas). Existing Traefik TCP health checks only verify port availability and will try to run the http healthcheck with an incorrect scheme, which is insufficient for routing exclusively to the active leader node or simply doesn't work.

Previously, setups had to rely on HAProxy or custom external health-checkers to bridge this gap. This PR allows Traefik to natively perform HTTP health checks on TCP backends, eliminating the need for external tools and keeping the entire proxying logic within Traefik.

What's changed
A existing healthCheck options allowed specifying an HTTP endpoint (host, port, path, status code) for a TCP service. They were just never used. Now we override the port, path and schema for the health check on this service

**Comment by eos175** [comment]
@supergoudvis116 

I completely agree\! The ability to run an **HTTP** health check on a **TCP** service is a brilliant and very practical solution. I'm really glad everything is working well.

For anyone who needs it, here's the usage example, which is quite straightforward:

```yaml
services:
  example_tcp_service:
    loadBalancer:
      servers:
        - address: 10.42.64.4:7445
        - address: 10.42.0.1:7445
        - address: 127.0.0.1:8448
      healthCheck:
        path: /health
        interval: "10s"
        timeout: "3s"
        mode: http # <-- The key is here
```

Thanks for sharing it; this greatly simplifies the configuration for scenarios like Patroni.

**Comment by supergoudvis116** [comment]
@nmengin 

Sorry for pinging you, could we get some input from you or someone else who can help putting this on the roadmap of traefik? We need some more input, I do feel like everyone is leaning more towards `Syn, Syn/Ack, Rst` `tcp-shaker` gets updated and PR's from @eos175 seem to be merged actively.

I am now using this on a dev stack and it just works beautifully but before I am even considering deploying this to staging and production I would like this functionality to be working from a released traefik version ;-).

This doesn't mean the PR of @eos175 needs to be merged per se, I am fine redoing this on someone else's branch again I just need the option to be able to run health checks on a TCP server.

Thank you!

**Comment by eos175** [comment]
Hi everyone, and thanks to @supergoudvis116  for reactivating this important discussion. I'd like to share my technical perspective on the best path forward, which has always been grounded in using the right tool for the right job.

From my point of view, Traefik should offer two clear and distinct modes for TCP service health checks:

1. The Pure TCP Check (mode: tcp): For verifying raw port availability, the `SYN, SYN-ACK, RST` handshake is the industry standard. It's lightweight, fast, and unambiguously answers the question: "Is a service listening on this port?". My original implementation in PR #10794 using `tcp-shaker` achieves exactly this. It's crucial to note that this approach does not introduce any new dependencies to the project, as `tcp-shaker` leverages libraries already present in Traefik's vendor tree.

2. The Application-Layer Check (`mode: http`): There are valid use cases, like the Patroni HA cluster example, where a simple port check is insufficient and an application-level response must be validated. The solution for this already exists within Traefik and it's brilliant: using `mode: http` on a TCP service's health check. This elegantly decouples the health check mechanism from the data plane routing, allowing the use of the powerful, existing HTTP checker to verify paths, status codes, etc., for any service that exposes an HTTP endpoint for its status.

Given these two clean and correct paths, I strongly believe that adding application-level logic (like payloads or initial handshakes) directly into the TCP health checker itself would be a step in the wrong direction. It creates a hybrid, non-standard mechanism that tries to replicate functionality we already have in a more robust and appropriate form.

I have been successfully running both of these configurations in my own environments. My recent contributions to [tcp-shaker](https://github.com/tevino/tcp-shaker/pull/31) were also aimed at ensuring the underlying tooling remains solid and maintainable.

@nmengin, to move this forward, I propose that we officially adopt this philosophy of two distinct, clear modes. I am fully available to help finalize this feature, whether that means reviving my original PR or collaborating on a new one that implements these well-defined principles. The goal for all of us is to provide the community with a robust, well-designed, and easy-to-maintain solution.

Thanks!!

**Comment by rtribotte** [comment]
Hi @eos175 @supergoudvis116,

Thanks for your work on this!
We would definitely like Traefik to provide three healthcheck mechanisms for TCP services.

The situation is a bit tedious. @ddtmachado has already opened https://github.com/traefik/traefik/pull/11238, and we would prefer to finish its review before adding anything else.
We haven't analyzed how your contributions differ, but since they are closely related, we think the work on #11238 should facilitate integrating your implementation later on, @eos175.

Would you be willing to help in that direction? (Helping reviewing @ddtmachado's PR)

**Comment by supergoudvis116** [comment]
Hi,

Yes I can do basically the same changes in his pr's if that speeds up the process. If @eos175 wants to help too that would be perfect!

**Comment by ddtmachado** [comment]
Hey, going for multiple optional mechanisms would be great. I rekon I haven't being updating my PR fast enough lately but I would be more than happy to help speed it up if needed. @eos175 / @supergoudvis116 fell free to branch from it or even submit PR's against my branch, I'll make sure I keep an eye on it.

@rtribotte / @sdelicata or any other maintainer, as usual, you're free to update my branch directly for any subjective topics (ie. config names, var names, field types etc) and if there is something more critical, like supporting new mechanisms or updated interfaces just keep me in the loop as I do like to keep up to date on those internals 📰

**Comment by supergoudvis116** [comment]
Hi @ddtmachado,

I was about to start adding the changes from @eos175 and myself into your PR, but I noticed that the building blocks are currently only set up for testing. Could I request a few changes?

At the moment, the healthchecker includes both HTTP and gRPC. I don’t think it makes much sense to separate the TCP logic on its own. Since I need some parts of the HTTP health check, I could either duplicate some logic into the ServiceTCPHealthChecker or consolidate them into the ServiceHealthChecker I feel like the latter approach makes more sense.

I’m not entirely sure which option is best—maybe @eos175 has some good ideas here—but I’d like to get things moving forward. ;-)

**Comment by nmengin** [comment]
Hello there,

Discussing this issue with the other maintainers, it appears that we were not all aligned on the way to proceed.
We would like to first merge @ddtmachado PR with its current scope, and, in a second step, bring the other modifications in a next dedicated PR.

@supergoudvis116 @eos175 Does it look good to you?

**Comment by supergoudvis116** [comment]
Definitely not ok for me, I think the decision to split up the TCP health check and gRPC/HTTP health check service makes no sense to me. The implementation is not near a final stage yet so I do feel like going with his PR over the one of @eos175 is in my honest opinion a weird decision. His implementation and my little add on for the http health check is already being used by us I don't mind putting that logic into his PR but waiting presumably extra months is a waste of time for everyone. I am fine doing a lot of the work myself but I really want this done in a more reasonable time frame. 

Hence I already was working on his PR but basically taking most of our changes but left his changes in case the decision would be made to prefer the separated logic for health checks. If there are valid arguments to use the other approach and people can explain why it would be beneficial I am down to do whatever it takes to get our changes to work but as of now I don't see any reason to.

**Comment by eos175** [comment]
Hello @nmengin, thank you for the clarification on the proposed path forward.

I understand the desire to merge things in sequential steps to maintain order. However, I share the architectural concerns raised by @supergoudvis116.

My main concern is that merging PR #11238 with its current architecture (a separate `TCPHealthChecker`) would force us to introduce a significant and breaking refactor in the very next PR. We would essentially be merging a design that we know we need to replace immediately.

From my perspective, a unified **`ServiceHealthChecker`** is the most robust long-term solution. Here's why:

1.  **Follows the DRY Principle (Don't Repeat Yourself):** It avoids code duplication, which is critical for long-term maintainability. More duplicated code means a larger surface for bugs and higher maintenance costs.
2.  **Enables Critical Features:** The recently discussed ability to run an **HTTP health check on a TCP service** (`mode: http`) is a perfect example. This feature inherently requires sharing logic with the HTTP checker, making a separate TCP checker architecturally unviable.

Regarding scope, the primary goal should be to integrate a pure TCP check via `tcp-shaker` (`SYN/ACK`). If the team later wants to add more complex checks, like inspecting N bytes or sending specific payloads, that could perfectly be handled as a **new, separate `mode` in the future**. The unified checker architecture provides the ideal foundation for this.

**My Proposal:**
Instead of a two-PR approach, I suggest we implement the unified architecture directly within the existing PR #11238.

I have been reviewing the latest changes in Traefik (up to `v3.5.0`), and I'm ready to help adapt my implementation to fit into @ddtmachado's branch following this unified structure. This would allow us to merge the right architecture from the start, saving everyone significant time and rework.

What are your thoughts on this approach? I believe it addresses your goal of moving forward while ensuring we build on a solid and maintainable foundation.

**Comment by rtribotte** [comment]
Hi @eos175,

Sorry for the late reply and for the misunderstanding!

> My Proposal:
Instead of a two-PR approach, I suggest we implement the unified architecture directly within the existing PR https://github.com/traefik/traefik/pull/11238.

Sure, that's what we were thinking. Architecture of #11238 should allow for the easy implementation of both TCP Health Check using SYN, SYN-ACK, and RST packets and HTTP Health Check.

We wanted to split the PRs by feature. However, we agree that merging the right architecture from the beginning is the best approach.

Thank you!

**Comment by SimonMTS** [comment]
Hey @rtribotte,

As an outside observer, there still seems to be a bit of a misunderstanding here. Please forgive my jumping it, but I'd also like TCP healthchecks :)

@eos175 and @supergoudvis116 are saying that the architecture in #11238 is not going to work, and will have to be replaced by the implementation from @eos175's fork anyway.

That doesn't match
>Architecture of #11238 should allow for the easy implementation of both TCP Health Check using SYN, SYN-ACK, and RST packets and HTTP Health Check.

It seems to me that @eos175's fork is basically ready to use (only missing the SYN/ACK flow?) while #11238 is not nearly as far along.

Wouldn't it make more sense to add SYN/ACK to @eos175's fork, rather than rewriting #11238 just to keep SYN/ACK?

**Comment by rtribotte** [comment]
Hi @SimonMTS,

Thanks for your feedback.

> @eos175 and @supergoudvis116 are saying that the architecture in https://github.com/traefik/traefik/pull/11238 is not going to work, and will have to be replaced by the implementation from @eos175's fork anyway.

> That doesn't match

I cannot confirm this yet, as I haven’t taken the time to properly compare both approaches.
But @eos175 suggested to use a unified architecture in #11238, which is what we are looking for.

> It seems to me that @eos175's fork is basically ready to use (only missing the SYN/ACK flow?) while https://github.com/traefik/traefik/pull/11238 is not nearly as far along.

The idea is to step in #11238 to improve it.
We also think that providing three health check approaches is relevant, and therefore don’t see why we should close that PR.

> My Proposal:
Instead of a two-PR approach, I suggest we implement the unified architecture directly within the existing PR https://github.com/traefik/traefik/pull/11238.

@eos175 Did I misunderstand your latest message?

**Comment by supergoudvis116** [comment]
> Hi [@SimonMTS](https://github.com/SimonMTS),
> 
> Thanks for your feedback.
> 
> > [@eos175](https://github.com/eos175) and [@supergoudvis116](https://github.com/supergoudvis116) are saying that the architecture in [#11238](https://github.com/traefik/traefik/pull/11238) is not going to work, and will have to be replaced by the implementation from [@eos175](https://github.com/eos175)'s fork anyway.
> 
> > That doesn't match
> 
> I cannot confirm this yet, as I haven’t taken the time to properly compare both approaches. But [@eos175](https://github.com/eos175) suggested to use a unified architecture in [#11238](https://github.com/traefik/traefik/pull/11238), which is what we are looking for.
> 
> > It seems to me that [@eos175](https://github.com/eos175)'s fork is basically ready to use (only missing the SYN/ACK flow?) while [#11238](https://github.com/traefik/traefik/pull/11238) is not nearly as far along.
> 
> The idea is to step in [#11238](https://github.com/traefik/traefik/pull/11238) to improve it. We also think that providing three health check approaches is relevant, and therefore don’t see why we should close that PR.
> 
> > My Proposal:
> > Instead of a two-PR approach, I suggest we implement the unified architecture directly within the existing PR [#11238](https://github.com/traefik/traefik/pull/11238).
> 
> [@eos175](https://github.com/eos175) Did I misunderstand your latest message?

I think he did the same as I did and only look into his changes more thoroughly when we wanted to actually integrate our changes. And just then came to the conclusion it doesn't work. I think I've should been more clear, my bad. 

In it's current state of pr #11238 is not working. He made tests but the actual implementation is lacking. We discussed wanting to have a unified `ServiceHealthChecker ` where his approach is making a new one just for his TCP healthcheck implementation. I am not bashing him but I think to get things going we should rethink our decision to go trough with #11238. I think the better decision should be integrate the way of how he wants to do the TCP health check into the work of @eos175. 

Can we all agree with this approach?

**Comment by ddtmachado** [comment]
@supergoudvis116 I opted for a separate healthcheck to follow the existing pattern (up to 3.x), as most features under the TCP stack had their own copy implementation from its regular http side. Also note that the PR was made with the intention of going in incremental steps, from the ground up, as I was not in a hurry and wanted to move it forward with iterative reviews from the maintainers, which is always nice because it triggers exactly this kind of design / implementation discussions and decisions must be taken along the way. 

It would be easier to just hack my way into the code and have a working implementation (which I did in a private branch) but I believe this iterative approach pays out in the long run.
 
@rtribotte if the decision is made to go with a common healthcheck package for both http and tcp I can definitely refactor the branch to accomodate that, I could also find some time for a live session to coordinate what is the direction and nail down any implementation details. I intentionally did not touch the PR any further while following the discussions here.

**Comment by supergoudvis116** [comment]
> [@supergoudvis116](https://github.com/supergoudvis116) I opted for a separate healthcheck to follow the existing pattern (up to 3.x), as most features under the TCP stack had their own copy implementation from its regular http side. Also note that the PR was made with the intention of going in incremental steps, from the ground up, as I was not in a hurry and wanted to move it forward with iterative reviews from the maintainers, which is always nice because it triggers exactly this kind of design / implementation discussions and decisions must be taken along the way.
> 
> It would be easier to just hack my way into the code and have a working implementation (which I did in a private branch) but I believe this iterative approach pays out in the long run.
> 
> [@rtribotte](https://github.com/rtribotte) if the decision is made to go with a common health check package for both HTTP and TCP I can definitely refactor the branch to accommodate that, I could also find some time for a live session to coordinate what is the direction and nail down any implementation details. I intentionally did not touch the PR any further while following the discussions here.

Yeah like I said i'm not bashing the decision nor hate on your work but I needed to get this going and it looked like your PR was stale. Making a unified http/tcp service health checker doesn't feel like a hack IMHO since it's purpose is for all services even if its from gRPC, HTTP and TCP. Currently HTTP and gRPC were there I think adding TCP there is thus not a hack. If you do have time please let me know how you want to proceed, I am down for whatever. Currently I am in favor of using @eos175 his implementation due to the fact I am already using it ;-). If there are good reasons to say that work is not the way to go due to future limitations let me know! 

If you want we can just chat over Discord or whatever to see if maybe we can come up with a plan which @rtribotte will support since you already have a history and may know what the roadmap of Traefik probably will look like.

**Comment by rtribotte** [comment]
Hi,

Thanks for all your feedback!

@eos175 The review of #12138 is now done.
The approach of duplicating the HeathChecker and adapt it for TCP, to us, looked to be the best way to go as of now.
The reason is not only how the healthcheck mechanism is configured and conducted, but also the need for using, on HTTP side, the roundTripper of the service, and on the TCP side, the dialer of the service.
Additionally, we considered the possibility of adding later a new TCP check mode (SYN, SYN-ACK, and RST packets), and nothing prevent that with this design.

@supergoudvis116 After second thoughts about you contribution to @eos175 branch, to allow HTTP healthcheck with TCP services, we think it worth opening a dedicated issue to track this need and know more about the community traction on this.
Could you please open a dedicated issue.
Design wise, we think the split approach doesn't prevent introducing such a feature.

**Comment by eos175** [comment]
Hello,

My apologies for the delayed response. I stopped receiving notifications for this thread and am just now catching up.

I see a decision has been made regarding the architecture moving forward. I understand completely.

Given that the decision is made, I will close my issue to avoid confusion.

Thanks everyone for your time.
